libnetwork/resolvconf/resolvconf.go

```diff
@@ -58,32 +66,478 @@ func Path() string {
 	return pathAfterSystemdDetection
 }
 
-const (
-	// ipLocalhost is a regex pattern for IPv4 or IPv6 loopback range.
-	ipLocalhost  = `((127\.([0-9]{1,3}\.){2}[0-9]{1,3})|(::1)$)`
-	ipv4NumBlock = `(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)`
-	ipv4Address  = `(` + ipv4NumBlock + `\.){3}` + ipv4NumBlock
-
-	// This is not an IPv6 address verifier as it will accept a super-set of IPv6, and also
-	// will *not match* IPv4-Embedded IPv6 Addresses (RFC6052), but that and other variants
-	// -- e.g. other link-local types -- either won't work in containers or are unnecessary.
-	// For readability and sufficiency for Docker purposes this seemed more reasonable than a
-	// 1000+ character regexp with exact and complete IPv6 validation
-	ipv6Address = `([0-9A-Fa-f]{0,4}:){2,7}([0-9A-Fa-f]{0,4})(%\w+)?`
+var (
+	defaultIPv4NSs = []string{"8.8.8.8", "8.8.4.4"}
+	defaultIPv6NSs = []string{"2001:4860:4860::8888", "2001:4860:4860::8844"}
 )
 
-var (
-	// Note: the default IPv4 & IPv6 resolvers are set to Google's Public DNS
-	defaultIPv4Dns = []string{"nameserver 8.8.8.8", "nameserver 8.8.4.4"}
-	defaultIPv6Dns = []string{"nameserver 2001:4860:4860::8888", "nameserver 2001:4860:4860::8844"}
-
-	localhostNSRegexp = regexp.MustCompile(`(?m)^nameserver\s+` + ipLocalhost + `\s*\n*`)
-	nsIPv6Regexp      = regexp.MustCompile(`(?m)^nameserver\s+` + ipv6Address + `\s*\n*`)
-	nsRegexp          = regexp.MustCompile(`^\s*nameserver\s*((` + ipv4Address + `)|(` + ipv6Address + `))\s*$`)
-	nsIPv6Regexpmatch = regexp.MustCompile(`^\s*nameserver\s*((` + ipv6Address + `))\s*$`)
-	nsIPv4Regexpmatch = regexp.MustCompile(`^\s*nameserver\s*((` + ipv4Address + `))\s*$`)
-	searchRegexp      = regexp.MustCompile(`^\s*search\s*(([^\s]+\s*)*)$`)
-	optionsRegexp     = regexp.MustCompile(`^\s*options\s*(([^\s]+\s*)*)$`)
+func defaultNSAddrs(ipv6 bool) []netip.Addr {
+	var addrs []netip.Addr
+	for _, a := range defaultIPv4NSs {
+		addrs = append(addrs, netip.MustParseAddr(a))
+	}
+	if ipv6 {
+		for _, a := range defaultIPv6NSs {
+			addrs = append(addrs, netip.MustParseAddr(a))
+		}
+	}
+	return addrs
+}
+
+// ResolvConf represents a resolv.conf file. It can be constructed by
+// reading a resolv.conf file, using method ReadResolvConfFile().
+type ResolvConf struct {
+	nameServers []netip.Addr
+	search      []string
+	options     []string
+	other       string // Unrecognised directives from the host's file, if any.
+
+	md metadata
```

Comment:
I did it that way because, while the getters are trivial, the setters aren't - in order to do the transforms (and print debug info into the generated file) the metadata needs to be kept in-line with what's happened.

Exporting the fields would make it easy for someone to break that.

Reply:
The metadata needs to be kept in sync with the transforms in order for the resolv.conf file to be generated _with debug info that accurately describes the transforms._ The metadata does not need to be in sync to generate a well-formed resolv.conf file.

_If_ `type ResolvConf` can be imported by code outside moby/moby and we have to maintain compatibility going forward, I want it to be broadly useful. That means not tightly coupling it to the sandbox_dns_unix use case; affording the generation of resolv.conf files from scratch or arbitrarily transforming a parsed resolv.conf file. That is easily accomplished by exporting the struct fields.

On the other hand, I don't care if the type is specialized to the sandbox_dns_unix use case if we are free to make breaking changes to `type ResolvConf` at any time.